### PR TITLE
Cache feature vectors to speed ranking

### PR DESCRIPTION
**Pull request by wyegelwel**

### CODE REVIEW

File: scheduler/src/cook/mesos/util.clj [hunk 1]
```diff
@@ -267,9 +267,22 @@
...
+   ;; This is done because accessing fields in datomic entities is much slower than
+   ;; a map access, even when accessing multiple times. 
+   ;; Don't want to complicate the function by caching new values in the event we see them
+   (let [task-ent->feature-vector (into {} 
+                                        (map #(vector % 
+                                                      (task->feature-vector %))) 
+                                        tasks)]
+     (fn [task1 task2]
+       (compare (or (task-ent->feature-vector task1)
+                    (task->feature-vector task1))
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/util.clj]: Why is the `or` necessary?

File: scheduler/src/cook/mesos/util.clj [hunk 2]
```diff
@@ -267,9 +267,22 @@
...
-  (fn [task1 task2]
-      (compare (task->feature-vector task1) (task->feature-vector task2))))
+  ([]
+   (same-user-task-comparator []))
+  ([tasks]
+   ;; Pre-compute the feature-vector for tasks we expect to see to improve performance
+   ;; This is done because accessing fields in datomic entities is much slower than
+   ;; a map access, even when accessing multiple times. 
+   ;; Don't want to complicate the function by caching new values in the event we see them
+   (let [task-ent->feature-vector (into {}
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/util.clj]: would `plumbing.core/map-from-keys` be a better fit here?

File: scheduler/src/cook/mesos/util.clj (hunk 1 above)
Comment (wyegelwel) [inline scheduler/src/cook/mesos/util.clj]: In the event that this comparator is called without pre-loading the tasks we still want it to work.

File: scheduler/src/cook/mesos/util.clj (hunk 2 above)
Comment (wyegelwel) [inline scheduler/src/cook/mesos/util.clj]: Yes, good point. Pushed a new version, it is much better thanks to your suggestion

File: scheduler/src/cook/mesos/util.clj (hunk 1 above)
Comment (shamsimam) [inline scheduler/src/cook/mesos/util.clj]: Just for my understanding, in the current context it is used (`sort-normal-jobs-by-dru`), it is not necessary, right?

**Comment by shamsimam** [comment]
lgtm

**Pull request merged**